lib/floe/workflow/state.rb

```diff
@@ -29,31 +29,53 @@ def initialize(workflow, name, payload)
         @comment  = payload["Comment"]
       end
 
-      def run!(input)
-        start(input)
-        sleep(1) while running?
+      def run_nonblock!
+        unless started?
+          context.step_nonblock_start
+          start(context.input)
+        end
+
+        return Errno::EAGAIN if running?
+
         finish
-        [context.next_state, context.output]
+
+        0
+      end
+
+      def run!(timeout: :forever)
+        start = Time.now.utc
+
+        while run_nonblock! == Errno::EAGAIN
+          return Errno::EAGAIN if timeout != :forever && (timeout.zero? || Time.now.utc - start > timeout)
+          sleep(1)
```

Comment:
Rubocop is going to complain about newline after return here
```suggestion
          return Errno::EAGAIN if timeout != :forever && (timeout.zero? || Time.now.utc - start > timeout)
          
          sleep(1)
```